### PR TITLE
Hpcc 16083 Provide ESP Binding information from WsESDLConfig

### PR DESCRIPTION
**Pull request by rpastrana**

### CODE REVIEW

**Comment by rpastrana** [comment]
@wangkx please review

File: esp/scm/ws_esdlconfig.ecm [hunk 1]
```diff
@@ -47,6 +47,17 @@ ESPresponse [exceptions_inline] PublishESDLDefinitionResponse
     ESPstruct BaseESDLStatus status;
 };
 
+ESPStruct MethodConfig
+{
+	String Name;
```
Comment (wangkx) [inline esp/scm/ws_esdlconfig.ecm]: indent 8?

File: esp/scm/ws_esdlconfig.ecm [hunk 2]
```diff
@@ -144,7 +146,7 @@ ESPresponse [exceptions_inline] GetESDLBindingResponse
     string EspProcName;   //Name of ESP Process
     string BindingName;
     string EspPort;
-    [depr_var("1.1")] String ConfigXML;
+    String ConfigXML;
```
Comment (wangkx) [inline esp/scm/ws_esdlconfig.ecm]: What if 1.1?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 3]
```diff
@@ -615,7 +616,28 @@ bool CWsESDLConfigEx::onPublishESDLBinding(IEspContext &context, IEspPublishESDL
...
+                methodsxml.appendf("<Method name='%s'", method.getName());
+                IArrayOf<IConstNamedValue> & attributes = method.getAttributes();
+                ForEachItemIn(attributesidx, attributes)
+                {
+                    IConstNamedValue& att = attributes.item(attributesidx);
+                    methodsxml.appendf(" %s='%s'", att.getName(), att.getValue());
+                }
+                methodsxml.append("/>");
+            }
+            methodsxml.append("</Methods>");
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: Do you need to check if (methods.ordinality() > 0) here?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 4]
```diff
@@ -615,7 +616,28 @@ bool CWsESDLConfigEx::onPublishESDLBinding(IEspContext &context, IEspPublishESDL
...
+            methodstree.setown(createPTreeFromXMLString(methodsxml.str()));
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: And here?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 5]
```diff
@@ -823,6 +843,21 @@ bool CWsESDLConfigEx::onConfigureESDLBindingMethod(IEspContext &context, IEspCon
 
         if (config.length() > 0)
             methodstree.setown(fetchConfigInfo(config, espProcName, espBindingName, esdlDefIdSTR, esdlServiceName));
+        else
+        {
+            StringBuffer methodsxml("<Methods>");
+
+            IConstMethodConfig& method = req.getMethodStructure();
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: What if no MethodStructure is set in req?

File: esp/scm/ws_esdlconfig.ecm [hunk 6]
```diff
@@ -85,6 +86,60 @@ ESPresponse [exceptions_inline] PublishESDLBindingResponse
     ESPstruct BaseESDLStatus status;
 };
 
+ESPrequest ListDESDLEspBindingsReq
+{
+	boolean IncludeESDLBindingInfo;
```
Comment (wangkx) [inline esp/scm/ws_esdlconfig.ecm]: indent 8?

File: esp/scm/ws_esdlconfig.ecm [hunk 7]
```diff
@@ -85,6 +86,60 @@ ESPresponse [exceptions_inline] PublishESDLBindingResponse
...
+    string BindingType;
+    string ServiceBuildSet;
+    string Port;
+    string Protocol;
+    ESPStruct ESDLBindingContents ESDLBinding;
+};
+
+ESPStruct ESPServerEx
+{
+	string Name;
```
Comment (wangkx) [inline esp/scm/ws_esdlconfig.ecm]: indent

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 8]
```diff
@@ -1347,6 +1348,116 @@ bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDe
     return true;
 }
 
+bool CWsESDLConfigEx::onListDESDLEspBindings(IEspContext &context, IEspListDESDLEspBindingsReq&req, IEspListDESDLEspBindingsResp &resp)
+{
+	bool includeESDLBindings = req.getIncludeESDLBindingInfo();
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: What if not set in req?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 9]
```diff
@@ -1347,6 +1348,116 @@ bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDe
     return true;
 }
 
+bool CWsESDLConfigEx::onListDESDLEspBindings(IEspContext &context, IEspListDESDLEspBindingsReq&req, IEspListDESDLEspBindingsResp &resp)
+{
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: Wrong indent inside this function

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 10]
```diff
@@ -1347,6 +1348,116 @@ bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDe
...
+							ForEach(*attributes)
+							{
+								Owned<IEspNamedValue> iespattribute = createNamedValue("","");
+								const char * attname = attributes->queryName()+1;
+								if (stricmp(attname, "name")==0)
+								{
+									methodconfig->setName(attributes->queryValue());
+								}
+								else
+								{
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: Move line 1411 here?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 11]
```diff
@@ -1347,6 +1348,116 @@ bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDe
...
+				desdlespbinding->setService(binding.getService());
+
+				if(includeESDLBindings) //this whole block should be in its own function
+				{
+					StringBuffer msg;
+					Owned<IPropertyTree> esdlbindingtree = getBindingTree(server.getName(), binding.getName(), msg);
+					if (esdlbindingtree)
+					{
+						Owned<IPropertyTree> def;
+						def.setown(esdlbindingtree->queryPropTree("Definition[1]"));
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: Do you need to check the 'def' before using it?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 12]
```diff
@@ -1347,6 +1348,116 @@ bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDe
...
+							{
+								Owned<IEspNamedValue> iespattribute = createNamedValue("","");
+								const char * attname = attributes->queryName()+1;
+								if (stricmp(attname, "name")==0)
+								{
+									methodconfig->setName(attributes->queryValue());
+								}
+								else
+								{
+									iespattribute->setName(attributes->queryName()+1);
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: You may use 'attname' here.

**Comment by wangkx** [comment]
@rpastrana looks good. A few questions.

File: esp/scm/ws_esdlconfig.ecm (hunk 2 above)
Comment (rpastrana) [inline esp/scm/ws_esdlconfig.ecm]: I should have never deprecated... that change was never released so we'll always return ConfigXML

**Comment by rpastrana** [comment]
@wangkx please review changes based on code review.

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 13]
```diff
@@ -1350,117 +1366,116 @@ bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDe
...
+                if(includeESDLBindings) //this whole block should be in its own function
+                {
+                    StringBuffer msg;
+                    Owned<IPropertyTree> esdlbindingtree = getBindingTree(server.getName(), binding.getName(), msg);
+                    if (esdlbindingtree)
+                    {
+                        Owned<IPropertyTree> def;
+                        def.setown(esdlbindingtree->queryPropTree("Definition[1]"));
+
+                        StringBuffer defid = def->queryProp("@id");
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: Is it possible that def is NULL?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 14]
```diff
@@ -1350,117 +1366,116 @@ bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDe
...
+                                    iespattributes.append(*iespattribute.getClear());
+                                }
+                            }
+                            methodconfig->setAttributes(iespattributes);
+                            iesmethods.append(*methodconfig.getClear());
+                        }
+
+                        msg.appendf("\nFetched ESDL Biding Configuration for %d methods.", iesmethods.length());
+
+						StringBuffer definition;
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: Wrong indent

**Comment by wangkx** [comment]
@rpastrana 2 comments

**Comment by rpastrana** [comment]
@wangkx one more review please

**Comment by wangkx** [comment]
@rpastrana looks fine.

**Comment by rpastrana** [comment]
@richardkchapman please merge.

**Pull request merged**